src/compiler/crystal/semantic/call.cr

```diff
@@ -361,7 +361,19 @@ class Crystal::Call
     end
   end
 
-  def instantiate(matches, owner, self_type, with_literals)
+  def instantiate(signature, matches, owner, self_type, with_literals)
+    if with_literals
+      # Now that we have all our matches, check if any of them matches exactly
+      # all types, assuming autocasted values will always match (because they
+      # matches and they were not ambiguous). If so, only keep matches up to
+      # that exact match. We need to do this here because with autocasting
+      # we consider all overloads to detect ambiguous usage.
+      stop_index = matches.index do |match|
+        signature.matches_exactly?(match, with_literals: true)
+      end
+      matches = matches[..stop_index] if stop_index
+    end
```

Comment:
I think this logic shouldn't go here. I'll try to put it somewhere else where it makes more sense.

Reply:
Or, maybe I'll leave a TODO to refactor it later...

Reply:
Actually, this is fine. Moving it to some other place will make it much complex. The `with_literals` logic is mostly kept in this file, so this might be a good place to do this check.

In general, the entire "with_literals" logic might benefit from a rewrite, but this PR is not the place to do it.